File: src/main/java/com/demigodsrpg/chitchat/PrivateMessage.java
package com.demigodsrpg.chitchat;

import com.google.gson.GsonBuilder;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class PrivateMessage implements Serializable {
    // -- SERIAL VERSION UID -- //

    private static final long serialVersionUID = 1L;

    // -- TRANSIENT DATA -- //

    private final transient Chitchat INST;

    // -- META DATA -- //

    private final String target;
    private final String sender;
    private final String message;

    // -- CONSTRUCTORS -- //

    @SuppressWarnings("unchecked")
    public PrivateMessage(Chitchat inst, String json) {
        INST = inst;
        Map<String, Object> map = new GsonBuilder().create().fromJson(json, Map.class);
        target = map.get("target").toString();
        sender = map.get("sender").toString();
        message = map.get("message").toString();
    }

    public PrivateMessage(Chitchat inst, String target, String sender, String message) {
        INST = inst;
        this.target = target;
        this.sender = sender;
        this.message = message;
    }

    // -- GETTERS -- //

    public String getTarget() {
        return target;
    }

    public String getSender() {
        return sender;
    }

    public String getMessage() {
        return message;
    }

    public Component getFormattedMessage(boolean out) {
       return Component.text("PM" + (out ? " to" : " from") + " <", NamedTextColor.DARK_GRAY).
                append(Component.text((out ? target : sender), NamedTextColor.DARK_AQUA)).
                append(Component.text(">: ", NamedTextColor.DARK_GRAY)).
                append(Component.text(message, NamedTextColor.GRAY));
    }

    public String getLogMessage() {
        return "PM <" + sender + " to " + target + ">: " + message;
    }

    public String toJson() {
        Map<String, Object> map = new HashMap<>();
        map.put("target", target);
        map.put("sender", sender);
        map.put("message", message);
        return new GsonBuilder().create().toJson(map, Map.class);
    }

    // -- SEND -- //

    public void send() {
        // Get the sender
        OfflinePlayer sender = Bukkit.getOfflinePlayer(this.sender);

        // Check if the player is on this server
        if (Bukkit.getPlayer(target) != null) {
            Bukkit.getPlayer(target).sendMessage(getFormattedMessage(false));
            Chitchat.getInst().getLogger().info(getLogMessage());
        } else if (Chitchat.getInst().USE_REDIS) {
            // Nope, send through redis
            RChitchat.REDIS_MSG.publishAsync(toJson());
        } else if (sender.isOnline()) {
            // Something went wrong
            Objects.requireNonNull(sender.getPlayer()).sendMessage(NamedTextColor.RED + "There was an error sending a private message.");
            return;
        }

        // Send the 'sent' message
        if (sender.isOnline()) {
            Objects.requireNonNull(sender.getPlayer()).sendMessage(getFormattedMessage(true));
        }

        // Add to the reply map
        INST.getReplyMap().put(this.sender, this.target);
        INST.getReplyMap().put(this.target, this.sender);
    }
}

File: src/main/kotlin/com/demigodsrpg/chitchat/PrivateMessage.kt
package com.demigodsrpg.chitchat

import com.google.gson.GsonBuilder
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import java.io.Serializable
import java.util.*

class PrivateMessage : Serializable {

    // -- TRANSIENT DATA -- //

    @Transient
    private lateinit var INST: Chitchat

    // -- META DATA -- //

    var target: String
    var sender: String
    var message: String

    // -- CONSTRUCTORS -- //

    constructor(json: String) {
        val map = GsonBuilder().create().fromJson(json, Map::class.java)
        target = map.get("target").toString()
        sender = map.get("sender").toString()
        message = map.get("message").toString()
    }

    constructor(target: String, sender: String, message: String) {
        this.target = target
        this.sender = sender
        this.message = message
    }

    // -- GETTERS -- //

    fun getFormattedMessage(out: Boolean): String {
        return ChatColor.DARK_GRAY.toString() + "PM" + (if (out) " to" else " from") + " <" + ChatColor.DARK_AQUA +
                (if (out) target else sender) + ChatColor.DARK_GRAY + ">: " + ChatColor.GRAY + message
    }

    fun getLogMessage(): String {
        return "PM <$sender to $target>: $message"
    }

    fun toJson(): String {
        val map = HashMap<String, Any>()
        map.put("target", target)
        map.put("sender", sender)
        map.put("message", message)
        return GsonBuilder().create().toJson(map, Map::class.java)
    }

    // -- SEND -- //

    fun send() {
        // Get the sender
        val sender = Bukkit.getOfflinePlayer(this.sender)

        // Check if the player is on this server
        if (Bukkit.getPlayer(target) != null) {
            Bukkit.getPlayer(target).sendMessage(getFormattedMessage(false))
            INST.logger.info(getLogMessage())
        } else if (INST.USE_REDIS) {
            // Nope, send through redis
            RChitchat.REDIS_MSG.publishAsync(toJson())
        } else if (sender.isOnline) {
            // Something went wrong
            sender.player.sendMessage(ChatColor.RED.toString() + "There was an error sending a private message.")
            return;
        }

        // Send the 'sent' message
        if (sender.isOnline) {
            sender.player.sendMessage(getFormattedMessage(true));
        }

        // Add to the reply map
        INST.REPLY_MAP.put(this.sender, this.target)
        INST.REPLY_MAP.put(this.target, this.sender)
    }
}

